### PR TITLE
Mutualize the pending auth handling (PSG-742)

### LINKED ISSUE #7193: Mutualize the pending auth handling
#### What would you like to do?
Mutualize the process of pending auth handling.

#### Why would you like to do it?
Several ViewModel which need authentication for some actions duplicate the same code to handle the pending auth and action.

#### How would you like to achieve it?
Create a dedicated `PendingAuthHandler` to encapsulate the logic and add unit tests on it.

### Have you considered any alternatives?

_No response_

### Additional context

_No response_

### Are you willing to provide a PR?

Yes

### PR DESCRIPTION
**Pull request by mnaturel**
## Type of change

- [ ] Feature
- [ ] Bugfix
- [x] Technical
- [ ] Other :

## Content

Creation of a new class `PendingAuthHandler` to mutualize the handling of pending auth for actions which require authentication. It was previously used in several `ViewModel`. I am doing this refactor to be able to test this and because we will need it for sign out action in the new device management screens.

## Motivation and context

Closes #7193 

## Tests

- Go to a screen using re-auth for example the device manager screen
- Go to settings -> Security & Privacy -> Show all sessions
- Long press on of your devices
- Press the sign out action
- Check the authentication flow is working well

## Tested devices

- [ ] Physical
- [x] Emulator
- OS version(s): Android 11

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/auth/PendingAuthHandler.kt [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+class PendingAuthHandler @Inject constructor(
+        private val matrix: Matrix,
+        private val activeSessionHolder: ActiveSessionHolder,
+) {
+
+    var uiaContinuation: Continuation<UIABaseAuth>? = null
+    var pendingAuth: UIABaseAuth? = null
+
+    fun ssoAuthDone() {
+        Timber.d("ssoAuthDone $pendingAuth , continuation: $uiaContinuation")
```
Comment (ganfra) [inline vector/src/main/java/im/vector/app/features/auth/PendingAuthHandler.kt:40]: I've not checked exactly the data in pendingAuth and continuation, but be careful not to log private data.

**Review by ganfra — APPROVED** [review]
Thanks for the refactor. Just look at my comment regarding log.

File: vector/src/main/java/im/vector/app/features/auth/PendingAuthHandler.kt (hunk 1 above)
Comment (mnaturel) [reply vector/src/main/java/im/vector/app/features/auth/PendingAuthHandler.kt:40]: Hum yes you are right. I reused an existing log but I will look into the data we could see with these models to be sure.

Comment (mnaturel) [reply vector/src/main/java/im/vector/app/features/auth/PendingAuthHandler.kt]: I removed the data from the logs as we may have sensitive data like in pendingAuth. See: [Updating the log in ssoAuthDone method to improve privacy](https://github.com/vector-im/element-android/pull/7206/commits/73e4c256ae23abc0fbf2fa1167f3126068ea2da1)

**Comment by bmarty** [comment]
Thanks for reducing code duplication, and adding tests!

**Pull request merged**